src/security/Handshake.cc

```diff
@@ -504,6 +521,54 @@ Security::HandshakeParser::parseSniExtension(const SBuf &extensionData) const
     return SBuf(); // SNI extension lacks host_name
 }
 
+/// RFC 8446 Section 4.2.1: SupportedVersions extension
+void
+Security::HandshakeParser::parseSupportedVersionsExtension(const SBuf &extensionData) const
+{
+    // RFC 8446 Section 4.1.2:
+    // In TLS 1.3, the client indicates its version preferences in the
+    // "supported_versions" extension (Section 4.2.1) and the legacy_version
+    // field MUST be set to 0x0303, which is the version number for TLS 1.2.
+    //
+    // RFC 8446 Section 4.2.1:
+    // A server which negotiates TLS 1.3 MUST respond by sending a
+    // "supported_versions" extension containing the selected version value
+    // (0x0304).  It MUST set the ServerHello.legacy_version field to 0x0303
+    // (TLS 1.2).
+    //
+    // To reduce misdetection of legacy agents and mishandling of malformed
+    // input, ignore supported_versions senders violating legacy_version MUSTs:
+    if (details->tlsSupportedVersion != AnyP::ProtocolVersion(AnyP::PROTO_TLS, 1, 2))
+        return;
+
+    // The code below ignores older protocol versions in supported_versions.
+    // TODO: Document why. For example, why leave tlsSupportedVersion at v1.2
+    // when a v1.3 server negotiates v1.1 using supported_versions?
```

Comment:
Please resolve this TODO.

Reply:
commit dfdf8e3f

@rousskov Please review the supported_versions extension parsing of client hello messages.
The RFC says about:
```
         Implementations of this specification MUST send this extension in the
         ClientHello containing all versions of TLS which they are prepared to
         negotiate (for this specification, that means minimally 0x0304, but
         if previous versions of TLS are allowed to be negotiated, they MUST
         be present as well).
```
My comment about is that to have a valid TLSv1.3 client hello message you should set legacy_version to v1.2 and include the v1.3 to supported_versions extension. If you have not a valid TLSv1.3 hello message can you have the supported_versions extension?

However the RFC 8446 section 4.2.1 also says:
 ```
    If this extension is present in the ClientHello, servers MUST NOT use
    the ClientHello.legacy_version value for version negotiation and MUST
    use only the "supported_versions" extension to determine client
    preferences. Servers MUST only select a version of TLS present in
    that extension and MUST ignore any unknown versions that are present    
    in that extension.  Note that this mechanism makes it possible to
    negotiate a version prior to TLS 1.2 if one side supports a sparse        
    range.  Implementations of TLS 1.3 which choose to support prior
    versions of TLS SHOULD support TLS 1.2.  Servers MUST be prepared to
    receive ClientHellos that include this extension but do not include
    0x0304 in the list of versions.
```
Probably we should not ignore older TLS versions in supported_versions for client hello messages. Opinion?

Reply:
```
this mechanism makes it possible to
negotiate a version prior to TLS 1.2 ...
Implementations of TLS 1.3 which choose to support prior
versions of TLS SHOULD support TLS 1.2.
```

TLS server Hello: Clearly, it is possible for a compliant TLS server to set selected_version to TLS v1.2 or even lower, indicating that as the final/negotiated protocol version. Thus, we cannot ignore any selected_version values like the current PR code does. We must honor the sent value instead (i.e. update tlsSupportedVersion, even if the update will lower that value). Please fix this bug.

As for TLS clients, I suggest adjusting the code to be explicit and consistent about handling the following two mutually exclusive cases:

1. A supported_versions extension that complies with all relevant RFC MUSTs. Implies legacy_version being TLS v1.2. Implies that there is at least one supported_versions entry indicating support for TLS v1.3 or above. We reset tlsSupportedVersion to the maximum supported_versions value and use that everywhere. There is a branch TODO in bio.cc about explaining or fixing one potential inconsistency, but otherwise we are done here.

2. All other cases. For simplicity, I suggest just honoring legacy_version here, even if it is TLS v1.3 or above. Honoring (somehow) protocol-violating supported_versions is, hopefully, unnecessary in this case. We can make this support more nuanced later if needed. To arrive at this simple solution, we need to adjust the code to check that there is at least one supported_versions entry indicating support for TLS v1.3 or above. That check becomes one of the preconditions to tlsSupportedVersion updates (per item 1).

Please let me know if you need more clarifications to make further progress with this issue.

Reply:
> TLS server Hello: Clearly, it is possible for a compliant TLS server to set selected_version to TLS v1.2 or even lower, indicating that as the final/negotiated protocol version. 

However in this case the TLS server must not send any supported_versions extension. From the RFC8446 section 4.2.1 section:
```
        A server which negotiates a version of TLS prior to TLS 1.3 MUST set
        ServerHello.version and MUST NOT send the "supported_versions"
        extension.
```
 The current strategy of this patch is to ignore the supported_versions extension in ServerHellos if it is shows TLSv1.2 or earlier and use legacy method to determine the selected TLS version. I think this is the correct.

> Thus, we cannot ignore any selected_version values like the current PR code does. We must honor the sent value instead (i.e. update tlsSupportedVersion, even if the update will lower that value). Please fix this bug.

@rousskov I am sure there is not any bug in the way this patch handles the TLS server hello. Please re-read my comment and the RFC 8442 section 4.2.1. However maybe we should be more strict  and reject (eg using Must clause): 
1. supported_version extension exists in a message with legacy_version!=TLSv1.2 2) 
2. supported_version extension with less than TLSv1.3 version in server hello message.

> There is a branch TODO in bio.cc about explaining or fixing one potential inconsistency

fixed in commit c2bcc0d0f

> As for TLS clients, I suggest adjusting the code to be explicit and consistent about handling the following two mutually exclusive cases:

My problem here is two conflicting paragraphs in RFC 8442 section 4.2.1, but probably I am missinterpreting them. Please check my commit 058113f2 for the fix I decided to make and the explanation here.

> 1. A supported_versions extension that complies with all relevant RFC MUSTs...

I think after commit 058113f2 this is ok now. Just the musts does not say that the TLSv1.3 must included in supportd_versions list for ClientHello messages


> 2. All other cases. For simplicity, I suggest just honoring legacy_version here, even if it is TLS v1.3 or above.

I think this is what the branch does.

Reply:
> However in this case the TLS server must not send any supported_versions extension. From the RFC8446 section 4.2.1 section:
> 
> ```
>         A server which negotiates a version of TLS prior to TLS 1.3 MUST set
>         ServerHello.version and MUST NOT send the "supported_versions"
>         extension.
> ```

I did not know that. Sorry if you told me about this MUST, and I forgot it. In this case, fromServer Hello parsing in 058113f2db6a8ecd00bc1d742b950e74c3345806 looks OK, but I will remove the "Ιf supported_versions included in server hello message the only valid value is the 0x0304" sentence because it is wrong and does not actually match the code -- values indicating "later" TLS versions are also valid (and we honor them). IMO, the already quoted RFC text alone is enough to explain fromServer Hello parsing code.

Somewhere, we need to document that all related MUST violations result in Squid using/trusting the legacy_version value. We almost do that already. I will polish this further.
 
> The current strategy of this patch is to ignore the supported_versions extension in ServerHellos if it is shows TLSv1.2 or earlier and use legacy method to determine the selected TLS version. I think this is the correct.

At least it is consistent with how we handle MUST violations in fromClient Hellos.

> However maybe we should be more strict and reject (eg using Must clause):

I do not think we should "reject" the whole handshake on a version MUST violation, but I may be wrong -- there are weak/handwaving arguments on both sides of this decision. If we do not discover a valid reason for such rejection, we will just wait for deployment experience to prove us wrong.


> > There is a branch TODO in bio.cc about explaining or fixing one potential inconsistency
> 
> fixed in commit [c2bcc0d](https://github.com/measurement-factory/squid/commit/c2bcc0d0f05b906ec93cab5495ecf37923281987)

Thank you.

> > As for TLS clients, I suggest adjusting the code to be explicit and consistent about handling the following two mutually exclusive cases:
 
> My problem here is two conflicting paragraphs in RFC 8442 section 4.2.1, but probably I am missinterpreting them.

Yes, you are probably ignoring the case of a compliant TLS v1.55 client implementation. The RFC warns TLS v1.3 and later server implementors that they `MUST be prepared to` handle that case. Otherwise, TLS protocol authors would have to invent a yet another version-negotiation extension when upgrading the TLS protocol beyond v1.3.

> Please check my commit [058113f](https://github.com/measurement-factory/squid/commit/058113f2db6a8ecd00bc1d742b950e74c3345806) for the fix I decided to make and the explanation here.

Looks good to me. Let me polish a couple of comments. The ball is on my side.

Reply:
My commit f31d214 polishes things a little but no functionality changes were intended. Please review that commit, but I am resolving this change request.